backend/src/services/secret-v2-bridge/secret-v2-bridge-service.ts

```diff
@@ -1038,38 +1188,128 @@ export const secretV2BridgeServiceFactory = ({
       });
     const folderId = folder.id;
 
-    const secretsToUpdate = await secretDAL.findBySecretKeys(
+    const secretsToUpdate = await secretDAL.find({
       folderId,
-      inputSecrets.map((el) => ({
-        key: el.secretKey,
-        type: SecretType.Shared
-      }))
-    );
+      $complex: {
+        operator: "and",
+        value: [
+          {
+            operator: "or",
+            value: inputSecrets.map((el) => ({
+              operator: "and",
+              value: [
+                {
+                  operator: "eq",
+                  field: "key",
+                  value: el.secretKey
+                },
+                {
+                  operator: "eq",
+                  field: "type",
+                  value: SecretType.Shared
+                }
+              ]
+            }))
+          }
+        ]
+      }
+    });
     if (secretsToUpdate.length !== inputSecrets.length)
       throw new NotFoundError({ message: `Secret does not exist: ${secretsToUpdate.map((el) => el.key).join(",")}` });
     const secretsToUpdateInDBGroupedByKey = groupBy(secretsToUpdate, (i) => i.key);
 
+    secretsToUpdate.forEach((el) => {
+      ForbiddenError.from(permission).throwUnlessCan(
+        ProjectPermissionActions.Edit,
+        subject(ProjectPermissionSub.Secrets, {
+          environment,
+          secretPath,
+          secretName: el.key,
+          secretTags: el.tags.map((i) => i.slug)
+        })
+      );
+    });
+
+    // get all tags
+    const sanitizedTagIds = inputSecrets.flatMap(({ tagIds = [] }) => tagIds);
+    const tags = sanitizedTagIds.length ? await secretTagDAL.findManyTagsById(projectId, sanitizedTagIds) : [];
+    if (tags.length !== sanitizedTagIds.length) throw new NotFoundError({ message: "Tag not found" });
```

Comment:
Maybe include which tag(s) are not found in message?

Reply:
Added found tags, For now because the provided is tag ids